keel-core/src/main/kotlin/com/netflix/spinnaker/keel/constraints/ArtifactTypeConstraintEvaluator.kt

```diff
@@ -0,0 +1,33 @@
+package com.netflix.spinnaker.keel.constraints
+
+import com.netflix.spinnaker.keel.api.DeliveryConfig
+import com.netflix.spinnaker.keel.api.Environment
+import com.netflix.spinnaker.keel.api.artifacts.ArtifactType
+import com.netflix.spinnaker.keel.api.artifacts.DeliveryArtifact
+import com.netflix.spinnaker.keel.core.api.ArtifactTypeConstraint
+import org.springframework.context.ApplicationEventPublisher
+import org.springframework.stereotype.Component
+
+@Component
+class ArtifactTypeConstraintEvaluator(
+  override val eventPublisher: ApplicationEventPublisher
+) : ConstraintEvaluator<ArtifactTypeConstraint> {
+  override val supportedType = SupportedConstraintType<ArtifactTypeConstraint>("artifact-type")
+  override fun isImplicit() = true
+
```

Comment:
The new constraint! It runs always to make sure that artifacts are only approved for the environment if they are the type that a resource uses.